Дополнительная задача

TatyanaV24 opened the pull request:
Дополнительную задачу вынесла в отдельную ветку и отдельный проект.
Задача решена не полностью.
Хочу получить Ваши комментарии, т.к. сейчас уперлась в стену и не понимаю, чего именно не замечаю.
С учетом условия задачи, рассуждаю так:
1. Пользователь наживает на кнопку "1", кнопка изменяет цвет и открывается первый контроллер.
2. Пользователь наживает на кнопку "2", кнопка изменяет цвет и открывается второй контроллер. При этом цвет кнопки "1" должен вернуться к первоначальному, а первый контроллер должен быть закрыт.
3. Пользователь наживает на кнопку "3", кнопка изменяет цвет и появляются два контрролера. При этом у кнопки "1" и "2" цвет сохраняется прежний, и являются закрытыми.
Мои кнопки меняют цвет и выводят контроллеры, но мне так и не удалось заставить скрывать нужный контроллер и возвращать кнопкам первоначальный цвет.

TatyanaV24 committed:
Дополнительная задача

Дополнительную задачу вынесла в отдельную ветку и отдельный проект.

ios-review commented on Bonus/Bonus/AppDelegate.swift:
```diff
@@ -0,0 +1,36 @@
+//
+//  AppDelegate.swift
+//  Bonus
```
Название ContainerViewController было бы более подходящим, так как в этом проекте разрабатывается контроллер типа контейнер.

TatyanaV24 replied:
Согласна, Ваш вариант лучше.

ios-review commented on Bonus/Bonus/ViewController.swift:
```diff
@@ -0,0 +1,69 @@
+
+class ViewController: UIViewController {
+    
+    let controlPanel = UIToolbar()
```
Очень круто, что поисследовали UIKit на наличие возможности отрисовать несколько кнопок.

ios-review commented on Bonus/Bonus/ViewController.swift:
```diff
@@ -0,0 +1,69 @@
+        addChild(child)
+        view.addSubview(child.view)
+        child.didMove(toParent: self)
+    }
```
👍

ios-review commented on Bonus/Bonus/ViewController.swift:
```diff
@@ -0,0 +1,69 @@
+        willMove(toParent: nil)
+        view.removeFromSuperview()
+        removeFromParent()
+    }
```
```
func remove(_ childVC: UIViewController) {
    willMove(toParent: nil)
    childVC.view.removeFromSuperview()
    childVC.removeFromParent()
}
```

ios-review commented on Bonus/Bonus/ViewController.swift:
```diff
@@ -0,0 +1,69 @@
+
+import UIKit
+
+class ViewController: UIViewController {
```
ContainerViewController управляют дочерними контент вью контроллерами, которые ему передаются извне.
Его задача только показывать/скрывать вью контроллеры, переданные ему клиентом этого контейнера.
По сути осуществлять навигацию.

Сейчас можно сказать, что данный вью контроллер просто Content View Controller, так как он полностью управляет всеми дочерними контроллерами.
Контентом дочерних контроллеров должен управлять тот, кто их создает, тот, кто использует класс ContainerViewController.

ContainerViewController может иметь следующий интерфейс:
```
class ContainerViewController: UIViewController {
    
    private var buttonsStackView: UIStackView
    private var childVCStackView: UIStackView
    
    private var childs: [UIViewController] = []
    private var placeholderVC: UIViewController?
    
    func addVC(_ vc: UIViewController, buttonTitle: String) {
        assert(childs.count < 6, "Too many child view controllers: only 6 allowed")
        // Сохраняем контроллер (но не показываем), создаем кнопку, показываем кнопку.
    }
    
    func setPlaceholder(_ vc: UIViewController) {
        // Сохраняем контроллер, который показывается в случае, если ни один контент контроллер не показан
    }
    
   @objc private func showHideContentVC(_ sender: UIButton) {
        // Показываем или скрываем контент контроллер, который соответствует кнопке
        // Если все контент контроллеры скрыты, то показываем placeholder
    }

    private func showChildVC(_ childVC: UIViewController) {
        // Функция для добавления контроллера в иерархию и его показа
    }
    
    private func hideChildVC(_ childVC: UIViewController) {
        // Функция для удаления контроллера из иерархии и его скрытия
    }
}
```
Как добавить или удалить контроллер из иерархии можно посмотреть по ссылке https://developer.apple.com/documentation/uikit/view_controllers/creating_a_custom_container_view_controller

Для настройки frame можно использовать autolayout с помощью констрейнтов и stack view.
Попробуйте совсем не пользоваться сторибордом и ксибами.

Сам контент контроллер может быть создан из кода, сконфигурирован с помощью других контент контоллеров извне, назначен как главный контроллер всего приложения.
Как создать UI из кода, не используя сториборд, описано по ссылке https://stackoverflow.com/questions/57963647/how-to-create-ui-programmatically-in-xcode-11
Смотреть нужно правильный ответ, который помечен зеленой галочкой.

В SceneDelegate, где должен создаваться ContainerViewController может быть следующий код для создания UI:

```
        let conVC = ContainerViewController()
        conVC.addVC(UIViewController(), buttonTitle: "one")
        conVC.addVC(UIViewController(), buttonTitle: "two")
        
        let defaultVC = UIViewController()
        defaultVC.view.backgroundColor = .lightGray
        conVC. setPlaceholder(defaultVC)
        
        window?.rootViewController = conVC
        window?.makeKeyAndVisible()
```

Предлагаю делать задание последовательно.
Вот первые задачи на выполнение:
1. Создать класс ContainerViewController и его интерфейс.
2. Создать экземпляр этого класса из кода.
3. Назначить его главным (rootViewController) для приложения также в коде.

После написать на скилбокс с ссылкой на этот ПР.
Будем постепенно разрабатывать этот контроллер.
Он довольно сложный и объемный.

ios-review reviewed (changes requested):
При нажатию на кнопку должен показываться или скрываться контроллер соответствующий только этой кнопке.
Состояние остальных должно быть неизменным.

В комментарии более подробные шаги для действий)

TatyanaV24 committed:
вопрос по контроллерам

У меня сейчас идут ошибки. Застопорилась на этом и дальше задачу не могу доделать.
Во-первых, интересует вот эта: 'ContainerViewController' cannot be constructed because it has no accessible initializers.
Получается мой класс не инициализирован, в SceneDelegate все данные внесены. Мне необходимо вводить дополнительную инициализацию, или я допустила ошибку и просто ее не вижу.
Также может от того, что долго на этой задаче сижу стала немного путаться.
Я правильно понимаю, цель задачи создать уникальный Container view controller, аналогично имеющемуся в системе Navigation Controller? Простите, что пишу слишком просто, пока выстраиваю все для себя на уровне "чайника", чтобы не только сделать, но и понимать, что именно сделала.
Дальше хочу разобрать логику того, что должно быть, также для моего понимания:
1. buttonsStackView и childVCStackView - необходимо использовать для настройки и компоновки кнопок и view.
2. childs и placeholderVC нужны для обработки контроллеров.
3. функция addVC должна сохранять контроллер и создавать и показывать кнопку. Кнопки надо создать другие, не тот вариант, который был использован мной?
4. функция setPlaceholder должна содержать контент контроллер, для отображения в случае если ни один не показыват.
5. showHideContentVC - скрывает или показывает контроллер в зависимости от того какая кнопка выбрана.

ios-review commented on Bonus/Bonus/ContainerViewController.swift:
```diff
@@ -0,0 +1,75 @@
+    let yellowViewController = YellowViewController()
+    
+    private var buttonsStackView: UIStackView
+    private var childVCStackView: UIStackView
```
Не опциональные свойства для класса необходимо инициализировать либо значениями по-умолчанию, либо с помощью инициализатора.
Чтобы избавиться от ошибок можно временно проинициализировать значениями по-умолчанию:
```
private lazy var buttonsStackView: UIStackView = UIStackView()
private lazy var childVCStackView: UIStackView = UIStackView()
```

ios-review commented on Bonus/Bonus/ContainerViewController.swift:
```diff
@@ -0,0 +1,75 @@
+class ContainerViewController: UIViewController {
+    let controlPanel = UIToolbar()
+    let greenViewController = GreenViewController()
+    let yellowViewController = YellowViewController()
```
Эти свойства удалить.

ios-review commented on Bonus/Bonus/ContainerViewController.swift:
```diff
@@ -0,0 +1,75 @@
+        twoBarButton.width = CGFloat(138)
+        threeBarButton.width = CGFloat(138)
+        controlPanel.setItems([oneBarButton, twoBarButton,threeBarButton], animated: false)
+        view.addSubview(controlPanel)
```
Реализация этого метода должна делать следующее:
1. Сохранять контроллер в массиве childs.
2. Создавать кнопку UIButton.
3. Настраивать кнопку (внешний вид, контент) из пункта 2.
4. Добавлять обработчик события нажатия на кнопку. 
https://developer.apple.com/documentation/uikit/uicontrol/1618259-addtarget
`#selector(showHideContentVC(sender:))`
5. Добавлять кнопку в `buttonsStackView`.
https://developer.apple.com/documentation/uikit/uistackview/1616227-addarrangedsubview

ios-review commented on Bonus/Bonus/ContainerViewController.swift:
```diff
@@ -0,0 +1,75 @@
+            showChildVC(yellowViewController)
+        default:
+            view.backgroundColor = .gray
+        }
```
Здесь будет более сложный алгоритм.

Предлагаю сперва полностью решить задания модуля 9, а затем вернуться сюда.

ios-review commented on Bonus/Bonus/SceneDelegate.swift:
```diff
@@ -0,0 +1,63 @@
+        // If using a storyboard, the `window` property will automatically be initialized and attached to the scene.
+        // This delegate does not imply the connecting scene or session are new (see `application:configurationForConnectingSceneSession` instead).
+        guard let windowScene = (scene as? UIWindowScene) else { return }
+        let window = UIWindow(windowScene: windowScene)
+        let conVC = ContainerViewController()
+        conVC.addVC(GreenViewController(), buttonTitle: "1")
+        conVC.addVC(YellowViewController(), buttonTitle: "2")
+        
+        let defaultVC = UIViewController()
+        defaultVC.view.backgroundColor = .lightGray
+        conVC.setPlaceholder(defaultVC)
+        
+        window.rootViewController = conVC
+        window.makeKeyAndVisible()
```
🔥

ios-review reviewed (changes requested):
Интерфейс готов к началу реализации алгоритмов!
Предлагаю перед реализацией немного переключиться и доделать задания модуля 9, чтобы полностью осознавать назначение двух UIStackView.

По вопросам:

> Получается мой класс не инициализирован, в SceneDelegate все данные внесены. Мне необходимо вводить дополнительную инициализацию, или я допустила ошибку и просто ее не вижу.

Ответил в комментарии.

> Я правильно понимаю, цель задачи создать уникальный Container view controller, аналогично имеющемуся в системе Navigation Controller?

Да, все так.
Задача заключается в реализации своего Container View Controller с определенными правилами навигации между Content View Controller.
До этого всегда разрабатывались только Content View Controller.

> 1. buttonsStackView и childVCStackView - необходимо использовать для настройки и компоновки кнопок и view.

UIStackView используется для простой верстки других вьюх.

> 2. childs и placeholderVC нужны для обработки контроллеров.

Container View Controller управляет показом/скрытием своих дочерних Content View Controller.
Content View Controller добавляются/удаляются в иерархию с помощью функцию showChildVC/hideChildVC.
Массив childs нужен, чтобы хранить все дочерние Content View Controller, которые могут быть показаны по нажатию на кнопку.
Это наше правило навигации для Container View Controller.
placeholderVC нужен, чтобы показать или скрыть его, в случае если все childs скрыты или хотя бы один показан соответственно.

> 3. функция addVC должна сохранять контроллер и создавать и показывать кнопку. Кнопки надо создать другие, не тот вариант, который был использован мной?

Ответил в комментарии.

> 4. функция setPlaceholder должна содержать контент контроллер, для отображения в случае если ни один не показыват.

Просто присвоить в свойство placeholderVC переданный в функцию контроллер.
Больше ничего.

> 5. showHideContentVC - скрывает или показывает контроллер в зависимости от того какая кнопка выбрана.

showHideContentVC - показывает/скрывает тот контроллер, который соответствует кнопке в зависимости от состояния кнопки (свойства isSelected).
Для первой кнопки - первый контроллер.
Для второй - второй.
И т.д.

Порядок определяется порядком вызова функции `func addVC(_ vc: UIViewController, buttonTitle: String)`
Та кнопка, которая будет создана в этой функции при вызове, будет соответствовать контроллеру, переданному в эту функцию.

TatyanaV24 committed:
исправление addVC

В соответсвии с Вашими комментариями внесла изменения в метод addVC.
Шла строго по этап и немного смутил пункт 2 (Создавать кнопку UIButton). Дело в том, что исходя из скрина, должно быть три кнопки, поэтому создала все три кнопки. Тем не менее не покидает ощущения, что должна быть создана только одна.
Не уверена в верной реализации метода setPlaceholder, пока это единственное что приходит в голову.
Жду Ваши замечания. Спасибо!

ios-review commented on Bonus/Bonus/SceneDelegate.swift:
```diff
@@ -0,0 +1,63 @@
+        // If using a storyboard, the `window` property will automatically be initialized and attached to the scene.
+        // This delegate does not imply the connecting scene or session are new (see `application:configurationForConnectingSceneSession` instead).
+        guard let windowScene = (scene as? UIWindowScene) else { return }
+        let window = UIWindow(windowScene: windowScene)
```
Данное окно (window) создается, показывается, но нигде не сохраняется.
Оно было показано и тут же удалено из памяти.

У объекта класса SceneDelegate есть свойство `var window: UIWindow?`.
Необходимо значение созданного окна присвоить в это свойство.

ios-review commented on Bonus/Bonus/ContainerViewController.swift:
```diff
@@ -0,0 +1,68 @@
+        assert(childs.count < 6, "Too many child view controllers: only 6 allowed")
+        // Сохраняем контроллер (но не показываем), создаем кнопку, показываем кнопку.
+        view.backgroundColor = UIColor.white
+        childs.append(GreenViewController())
+        childs.append(YellowViewController())
+        let oneButton = UIButton()
+        let twoButton = UIButton()
+        let threeButton = UIButton()
+        oneButton.frame = CGRect(x: 0, y: 40, width: 138, height: 40)
+        oneButton.backgroundColor = UIColor.gray
+        oneButton.setTitle("1",for: .normal)
+        oneButton.addTarget(self, action: #selector(showHideContentVC(sender:)), for: .valueChanged)
+        twoButton.frame = CGRect(x: 138, y: 40, width: 138, height: 40)
+        twoButton.backgroundColor = UIColor.gray
+        twoButton.setTitle("2",for: .normal)
+        twoButton.addTarget(self, action: #selector(showHideContentVC(sender:)), for: .valueChanged)
+        threeButton.frame = CGRect(x: 276, y: 40, width: 138, height: 40)
+        threeButton.backgroundColor = UIColor.gray
+        threeButton.setTitle("3",for: .normal)
+        threeButton.addTarget(self, action: #selector(showHideContentVC(sender:)), for: .valueChanged)
+        buttonsStackView.addArrangedSubview(oneButton)
+        buttonsStackView.addArrangedSubview(twoButton)
+        buttonsStackView.addArrangedSubview(threeButton)
+        view.addSubview(buttonsStackView)
```
Данная функция должна работать только с входными параметрами.
А именно сохранять в массиве childs переданный в функцию в параметре vc: UIViewController контроллер.
И создавать только одну кнопку, которая будет соответствовать переданному UIViewController, на основе параметра buttonTitle.
Эта кнопка должна быть добавлена в buttonsStackView.
С помощью этой функции клиент этого контейнера сможет сконфигурировать его теми контент контроллерами, какими хочет.
Клиент - тот программист, который будет использовать класс для решения своих задач.

Сам buttonsStackView и childVCStackView могут быть добавлены в иерархию вью в методе `viewDidLoad()`

ios-review reviewed (changes requested):
Над функцией addVC надо еще поработать.
Объект window необходимо сохранять в свойстве, чтобы окно не удалялось сразу.
